doc/ccze-cssdump.1.in
.\" Hey Emacs! This is -*- nroff -*-!
.\" Copyright (C) 2003 Gergely Nagy <algernon@@bonehunter.rulez.org>
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one.
.\"
.\" Permission is granted to copy and distribute translations of this
.\" manual into another language, under the above conditions for modified
.\" versions, except that this permission notice may be stated in a
.\" translation approved by the Author.
.TH CCZE-CSSDUMP 1 "2003-01-13" "CCZE @VERSION@" CCZE
.SH NAME
ccze\-cssdump \- Dump CCZE color setup into CSS format
.SH SYNOPSIS
.B ccze\-cssdump
.RI [ options ]
.SH DESCRIPTION
This manual page documents briefly the
.B ccze\-cssdump
utility, which is a simple tool to dump the color setup of CCZE into
Cascading Style Sheet format, to be used later by \fBccze \-h \-o
cssfile\fR.
.SH OPTIONS
This program follows the usual GNU command line syntax, with long
options starting with two dashes (`\-'). A summary of options is
included below.
.TP
.B \-F, \-\-rcfile \fIrcfile\fB
Read \fIrcfile\fR as a configuration file upon startup.
.TP
.B \-l, \-\-load
Load the default configuration files upon startup.
.TP
.B \-\-help
Show summary of options and exit.
.TP
.B \-V, \-\-version
Show version of program.
.SH FILES
.TP
.IR @sysconfdir@/colorizerc ", " $HOME/.colorizerc
.RS
These files are the default configuration files for \fBcolorize\fR,
and are parsed by \fBccze\fR for the sake of full compatibility.
.RE
.IR @sysconfdir@/cczerc ", " $HOME/.cczerc
.RS
This two are the main configuration files, in which one can change the
colors used by the program to his liking. See the comments in the
beginning of \fI@sysconfdir@/cczerc\fR for a description on the files'
structure.
.RE
.SH "SEE ALSO"
.BR ccze (1)
.SH AUTHOR
ccze was written by Gergely Nagy <algernon@bonehunter.rulez.org>,
based on colorize by Istvan Karaszi <colorize@spam.raszi.hu>.


NEWS
NEWS -- list of user-visible changes		-*- outline -*-

* CCZE 0.2.1 [Fireflower] -- 2003-05-13

The Fireflower release brings transparency and background colour
support for the raw ANSI output mode, documentation updates and minor
bugfixes all over the place.

* CCZE 0.2.0 [Mephisto] -- 2003-03-29

Mephisto features raw ANSI output support, without the use of
nCurses. This adds support for scrollback, both on console and in
xterms. Also, it allows one to send CCZE to the background, and mix
various logs and options on one terminal.

A new coloriser for Pure-FTPD's Stats log format was added, and the
httpd module was modified to handle Pure-FTPD's CLF format, and bad
HTTP requests.

* CCZE 0.1.228 [Largo] -- 2003-03-20

This release features many many internal enhancements, mostly for the
plug-in system, a coloriser for proftpd logs, and various fixes all
over the place.

* CCZE 0.1.212 [Avatar] -- 2003-03-04

Implemented support for passing arguments to plugins. None of the
plugins shipped with CCZE support this, however, external plugins
might (and do, for that matter, see the homepage).

Also, some minor bugs (misformatting on the manual page, garbled
output, and not properly recognised procmail logs) were also squashed.

* CCZE 0.1.200 [Enigma of the Absolute] -- 2003-02-21

During the past few weeks, CCZE was ported to Cygwin and Tru64 (many
thanks to Jon Ewing and Erik Braun for their help).

Also, a coloriser for distccd.log files was added too.

* CCZE 0.1.171 [Mesmerism] -- 2003-02-09.

This release brings colorisers for php.log, oops/oops.log,
icecast/icecast.log and icecast/usage.log. Also, it is now possible to
disable any of the built-in plugins, that is, they behave as loadable
modules when --plugin is specified.

For the benefit of transparent xterm users, CCZE defaults to not alter
the background colour by default.

* CCZE 0.1.150 [The Cardinal Sin] -- 2003-02-04.

Since the last release, a great deal of code cleanup took place,
resulting in most of the modules' sources getting renamed. As usual,
bugs were fixed during this process, so certain characters will get
escaped in the HTML output.

Also, a full-blown coloriser module was added for xferlog files, and a
syslog-subcoloriser for postfix logs.

CCZE also handles SIGHUP nicely, by effectively restarting itself.

* CCZE 0.1.134 [Song of the Dispossessed] -- 2003-01-19.

Colours (even the background of the outputted HTML) can now be changed
on the command-line too. During this process, CCZE was ported to
FreeBSD.

* CCZE 0.1.122 [Wild in The Woods] -- 2003-01-13.

HTML output was cleaned up and made customisable. From now on, CCZE
uses Cascading Style Sheets.

* CCZE 0.1.108 [Devorzhum] -- 2003-01-10.

In this release experimental HTML output was added. It is not really
customisable yet, but produces usable output (better than that of
colorize! :).

* CCZE 0.1.103 [For My Fallen Angel] -- 2002.12.02.

Many internal cleanups and documentation updates, support for built-in
and partial plugins, and more.

* CCZE 0.1.91 [Like Gods of the Sun] -- 2002.11.19.

Major restructuring in the plugin code: cleaner, better, lighter. The
new design allows sub-colorisers, to colorise syslog messages (with
the date, host and that kind of stuff handled by the syslog module)
for example. Two such plugins are included: one for ulogd, another for
apmd.

* CCZE 0.1.80 [The Stance of Evander Sinque] -- 2002.10.12.

This is mainly a minor convenience-enhanced release. The major change
is that CCZE will now exit if it could not load any plugins.

* CCZE 0.1.72 [Base Level Erotica] -- 2002.10.04.

Brown paper-bag release, as I broke all the modules in 0.1.64. Apart
from this fix, this release is the same as 0.1.70 yesterday.

* CCZE 0.1.70 [Edenbeast] -- 2002.10.03.

Initial public release.


src/ccze-cssdump.c
/* -*- mode: c; c-file-style: "gnu" -*-
 * ccze-cssdump.c -- Dump internal color table into css format
 * Copyright (C) 2003 Gergely Nagy <algernon@bonehunter.rulez.org>
 *
 * This file is part of ccze.
 *
 * ccze is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * ccze is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 * License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#include <ccze.h>
#ifdef HAVE_ARGP_H
# include <argp.h>
#endif

#include "ccze-compat.h"
#include "ccze-private.h"
#include "ccze-color.c"

const char *argp_program_name = "ccze-cssdump";
const char *argp_program_version = "ccze-cssdump (ccze 0.1." PATCHLEVEL ")";
const char *argp_program_bug_address = "<algernon@bonehunter.rulez.org>";
static struct argp_option options[] = {
  {NULL, 0, NULL, 0, "", 1},
  {"rcfile", 'F', "FILE", 0, "Read configuration from FILE", 1},
  {"load", 'l', NULL, 0, "Load default configuration files", 1},
  {NULL, 0, NULL, 0,  NULL, 0}
};

ccze_config_t ccze_config = {
  .scroll = 1,
  .convdate = 0,
  .remfac = 0,
  .wcol = 1,
  .slookup = 1,
  .rcfile = NULL,
  .cssfile = NULL,
  .transparent = 1,
  .pluginlist_len = 0,
  .pluginlist_alloc = 10,
  .color_argv_len = 0,
  .color_argv_alloc = 10,
  .mode = CCZE_MODE_HTML
};

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static struct argp argp =
  {options, parse_opt, 0, "ccze -- cheer up 'yer logs.", NULL, NULL, NULL};

static int ccze_loaddefs = 0;

static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
  switch (key)
    {
    case 'F':
      ccze_config.rcfile = arg;
      break;
    case 'l':
      ccze_loaddefs = 1;
      break;
    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}

int
main (int argc, char *argv[])
{
  argp_parse (&argp, argc, argv, 0, 0, NULL);
  
  ccze_color_init ();

  if (ccze_config.rcfile)
    ccze_color_load (ccze_config.rcfile);
  else if (ccze_loaddefs)
    {
      char *home, *homerc;
      
      ccze_color_load (SYSCONFDIR "/colorizerc");
      ccze_color_load (SYSCONFDIR "/cczerc");
      home = getenv ("HOME");
      if (home)
	{
	  asprintf (&homerc, "%s/.colorizerc", home);
	  ccze_color_load (homerc);
	  free (homerc);
	  asprintf (&homerc, "%s/.cczerc", home);
	  ccze_color_load (homerc);
	  free (homerc);
	}
    }

  ccze_colors_to_css ();
  return 0;
}
